### DISCUSSION TITLE
Allow to specify a configuration folder location instead of the default ~/.sfdx or %USERPROFILE%\.sfdx

### DISCUSSION THREAD

**Discussion by mokchend** [forcedotcom/cli#2700] [Ideas]
This repository only supports the Salesforce CLI Core libraries if you have a feature request that is not related to the Core Libraries but instead to the Salesforce CLI or the force commands please submit your issue to the forcedotcome/cli repo not this one. 

**Is your feature request related to a problem? Please describe.**
I'm always frustrated when once I have setup my sfdx aliases, and switch my OS: in docker, in WSL2, in VM ... I lost all conf DX config & alias as i'm not able to specify my own (shared) .sfdx directory amongs differents OS.
Please provide a parameter (defined as optionnal) to specify our own .sfdx directory location.

**Describe the solution you'd like**
Require the ability to specify a configuration folder that handle the DX (hidden) folder, by default on your user home directory called .sfdx

On Mac OS and Linux, you can find your DX config folder in ~/.sfdx
On Windows, you can use %USERPROFILE%\.sfdx

**Describe alternatives you've considered**
This option is usefull when you want to share your SFDX config folder betwen multiple OS, docker machine or WSL2

**Additional context**
N/A

**Comment by Exordian** [comment]
Please also move from ~/.sfdx to the XDG spec, a library for that would be [1]. This will also make the config directory configurable.

[1] https://github.com/folder/xdg

**Comment by R0Wi** [comment]
If we would for example introduce a new environment variable `SFDX_DIR` (or the newer `SF_DIR`), one could allow the end user to overwrite the global sfdx/sf state folders by adjusting [these lines](https://github.com/forcedotcom/sfdx-core/blob/d426be78632d2cc8cef4531fb2e8b50d667b7174/src/global.ts#L58-L60) to something like:

```ts
  public static get SFDX_DIR(): string {
    const baseDir = env.getString('SFDX_DIR', os.homedir());
    return path.join(baseDir, Global.SFDX_STATE_FOLDER);
  }
```

@WillieRuemmele is there any interest in incorporating such a feature?

**Comment by ragoergens** [comment]
@R0Wi I'd like to share my .sdfx folder between Windows and WSL2.
- Summary: symlinking fails due to a permissions issue, fallback to copy Windows -> Linux.
- The challenge for a shared directory between different OSs/VMs are the different security mechanisms/requirements

### Experimentation

- If I just create a symlink, ~/.sfdx to /mnt/c/Users/%username%/.sfdx
    - The symlink is being followed but `sf org list` says, for all connections:
    Warning: The auth file for rene.goergens@[...] is invalid.
    [...] 
    No Orgs found
- If I manually create an empty ~/.sfdx and do cp /mnt/c/Users/%username%/.sfdx/* .sfdx, the result is the same:
    Warning: The auth file for rene.goergens@[...] is invalid.
    [...] 
    No Orgs found
- If I use sf to create a connection under WSL2 so that ~/.sfdx gets created by sf, suddenly things are looking better:
    - Replacing the key in key.json with the one from /mnt/c/Users/%username%/.sfdx/key.json and copying the rest of the files works!
    - The gotcha: under Linux, sf requires the correct file permissions for key.json, otherwise it will say "The auth file for ... is invalid."
    The correct permissions are: -rw------- (chmod 600, as created under Linux), not -rwxr-xr-x (chmod 755, as copied from Windows)

### My solution

- So to use Windows/Linux in a master/slave configuration, upon WSL2 launch I want to:
    - Flush the Linux side: rm ~/.sfdx/*
    - Copy from Windows: cp /mnt/c/Users/%username%/.sfdx/* ~/.sfdx
    - Set the correct permissions: chmod 600 ~/.sfdx/key.json
- So I put the following in my ~/.profile (replace %username% with actual username):

```
# flush .sfdx and copy connections from Windows
if [ -d "$HOME/.sfdx" ] ; then
    rm ~/.sfdx/*
    cp /mnt/c/Users/%username%/.sfdx/* ~/.sfdx
    chmod 600 ~/.sfdx/key.json
fi
```

**Comment by cwelsys** [comment]
I'd like to add my voice in support of implementing XDG spec, or at least allowing us to the define a SF_CONFIG_DIR.

#2819 mentions release notes reference putting files in $HOME/.config/sf, I cannot find any reference to environment variables that enable this behavior. Was this scrapped?

**Comment by cwelsys** [comment]
I've put together an implementation for this and opened a PR: https://github.com/forcedotcom/sfdx-core/pull/1266

The change introduces `SF_HOME` and `SFDX_HOME` environment variables that override where sfdx-core resolves the global `.sf` and `.sfdx` directories. This is essentially what @R0Wi proposed above, extended to cover all the code paths that resolve global state directories.

If there is any interest from the maintainers in pursuing this, I can draft the needed changes to the downstream packages that also hardcode these paths:   

- [forcedotcom/salesforcedx-templates](https://github.com/forcedotcom/salesforcedx-templates) `gitRepoUtils` constructs `.sfdx` paths via `os.homedir()` directly
- [salesforcecli/cli-plugins-testkit](https://github.com/salesforcecli/cli-plugins-testkit) `hubAuth` reads auth files from a hardcoded `.sfdx` path
